core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpRequest.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.SplitHttpRequest;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+
+import io.netty.util.concurrent.EventExecutor;
+
+public class DefaultSplitHttpRequest extends AbstractSplitHttpMessage implements SplitHttpRequest {
+
+    private static final HttpHeaders EMPTY_TRAILERS = HttpHeaders.of();
+
+    private final BodySubscriber bodySubscriber = new SplitHttpRequestBodySubscriber();
+    private final HttpRequest request;
+
+    public DefaultSplitHttpRequest(HttpRequest request, EventExecutor executor) {
+        super(request, executor);
+        this.request = request;
+        request.subscribe(bodySubscriber, upstreamExecutor, SubscriptionOption.values());
+    }
+
+    @Override
+    public RequestHeaders headers() {
+        return request.headers();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        subscribe0(subscriber, bodySubscriber, executor, options);
+    }
+
+    private final class SplitHttpRequestBodySubscriber extends BodySubscriber {
```

Comment:
Is it possible to make `BodySubscriber` a concrete class by taking some constructor parameters?
```java
BodySubscriber(int prepatch, @Nullable BiConsumer<ResponseHeaders, Subscription> onResponseHeaders) {
   ...
}
```

Reply:
@ikhoon, Are you referring to the fix included in the comment below...??
- https://github.com/line/armeria/pull/3953/commits/bc8550e76b6b74cb8999243f1aa9ee7aa7a1eacc

Reply:
Had a chat with @ikhoon and it's a bit hard to pass the response specific functions to constructor because we need two separate functions: Setting `ResponseHeaders` in `onNext` and Setting `ResponseHeaders` in `maybeCompleteHeaders`.

Rather, we can just override the `BodySubscriber` in `DefaultSplitHttpResponse`.

```java
public class DefaultSplitHttpResponse extends AbstractSplitHttpMessage implements SplitHttpResponse {

    private static final ResponseHeaders HEADERS_WITH_UNKNOWN_STATUS = ResponseHeaders.of(HttpStatus.UNKNOWN);

    private final SplitHttpResponseBodySubscriber bodySubscriber;

    public DefaultSplitHttpResponse(HttpResponse response, EventExecutor executor) {
        this(response, executor, new SplitHttpResponseBodySubscriber(response, executor));
    }

    private DefaultSplitHttpResponse(HttpResponse response, EventExecutor executor,
                                     SplitHttpResponseBodySubscriber bodySubscriber) {
        super(response, executor, bodySubscriber);
        this.bodySubscriber = bodySubscriber;
    }

    @Override
    public final CompletableFuture<ResponseHeaders> headers() {
        return bodySubscriber.headersFuture();
    }

    private static class SplitHttpResponseBodySubscriber extends BodySubscriber {
        private final HeadersFuture<ResponseHeaders> headersFuture = new HeadersFuture<>();

        SplitHttpResponseBodySubscriber(HttpResponse response, EventExecutor executor) {
            super(1, response, executor);
        }

        CompletableFuture<ResponseHeaders> headersFuture() {
            return headersFuture;
        }

        @Override
        public void onNext(HttpObject httpObject) {
            if (httpObject instanceof ResponseHeaders) 
                ...
                return;
            }
            super.onNext(httpObject);
        }

        @Override
        void maybeCompleteHeaders(@Nullable Throwable cause) {
            ...
            super.maybeCompleteHeaders(cause);
        }
    }
}
```

In `DefaultSplitHttpRequest` we can just pass the `BodySubscriber`:
```java
public class DefaultSplitHttpRequest extends AbstractSplitHttpMessage implements SplitHttpRequest {

    private final RequestHeaders headers;

    public DefaultSplitHttpRequest(HttpRequest request, EventExecutor executor) {
        super(request, executor, new BodySubscriber(0, request, executor));
        headers = request.headers();
    }

    @Override
    public RequestHeaders headers() {
        return headers;
    }
}
```

Also, we might want to move `BodySubscriber` out of `AbstractSplitHttpMessage` so that we can simplify `AbstractSplitHttpMessage` and encapsulate everything in `BodySubscriber`:
```java
abstract class AbstractSplitHttpMessage implements SplitHttpMessage, StreamMessage<HttpData> {

    @SuppressWarnings("rawtypes")
    private static final AtomicReferenceFieldUpdater<BodySubscriber, Subscriber>
            downstreamUpdater = AtomicReferenceFieldUpdater.newUpdater(BodySubscriber.class, Subscriber.class,
                                                                       "downstream");

    private final HttpMessage upstream;
    protected final EventExecutor upstreamExecutor;
    private final BodySubscriber bodySubscriber;

    protected AbstractSplitHttpMessage(HttpMessage upstream, EventExecutor executor, BodySubscriber bodySubscriber) {
        this.upstream = requireNonNull(upstream, "upstream");
        upstreamExecutor = requireNonNull(executor, "executor");
        this.bodySubscriber = bodySubscriber;
        upstream.subscribe(bodySubscriber, upstreamExecutor, SubscriptionOption.values());
    }

    @Override
    public final StreamMessage<HttpData> body() {
        return this;
    }

    @Override
    public final CompletableFuture<HttpHeaders> trailers() {
        return bodySubscriber.trailersFuture();
    }
    ...

    @Override
    public final boolean isEmpty() {
        return !isOpen() && !bodySubscriber.wroteAny();
    }

    ...

    @Override
    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
                          SubscriptionOption... options) {
        requireNonNull(subscriber, "subscriber");
        requireNonNull(executor, "executor");
        requireNonNull(options, "options");

        if (!downstreamUpdater.compareAndSet(bodySubscriber, null, subscriber)) {
            subscriber.onSubscribe(NoopSubscription.get());
            subscriber.onError(new IllegalStateException("subscribed by other subscriber already"));
            return;
        }

        if (executor.inEventLoop()) {
            bodySubscriber.initDownstream(subscriber, executor, options);
        } else {
            executor.execute(() -> bodySubscriber.initDownstream(subscriber, executor, options));
        }
    }
}
```
```java
class BodySubscriber implements Subscriber<HttpObject>, Subscription {
    ...
    protected BodySubscriber(int prefetch, HttpMessage upstreamMessage, EventExecutor upstreamExecutor) {
        pendingRequests = prefetch;
        this.upstreamMessage = upstreamMessage;
        this.upstreamExecutor = upstreamExecutor;
    }
    ...
}
```
WDYT?